src/bounded_range.rs

```diff
@@ -0,0 +1,602 @@
+
+            // Complement of ∞ is ∅
+            Some((Unbounded, Unbounded)) => Self::none(),
+
+            // First high bound is +∞
+            Some((Included(v), Unbounded)) => Self::strictly_lower_than(v.clone()),
+            Some((Excluded(v), Unbounded)) => Self::lower_than(v.clone()),
+
+            Some((Unbounded, Included(v))) => {
+                Self::negate_segments(Excluded(v.clone()), &self.segments[1..])
+            }
+            Some((Unbounded, Excluded(v))) => {
+                Self::negate_segments(Included(v.clone()), &self.segments[1..])
+            }
+            Some((Included(_), Included(_)))
+            | Some((Included(_), Excluded(_)))
+            | Some((Excluded(_), Included(_)))
+            | Some((Excluded(_), Excluded(_))) => Self::negate_segments(Unbounded, &self.segments),
+        }
+    }
+
+    /// Helper function performing the negation of intervals in segments.
+    fn negate_segments(start: Bound<V>, segments: &[Interval<V>]) -> Self {
+        let mut complement_segments: SmallVec<Interval<V>> = SmallVec::empty();
+        let mut start = start;
+        for (v1, v2) in segments {
+            complement_segments.push((
+                start,
+                match v1 {
+                    Included(v) => Excluded(v.clone()),
+                    Excluded(v) => Included(v.clone()),
+                    Unbounded => unreachable!(),
+                },
+            ));
+            start = match v2 {
+                Included(v) => Excluded(v.clone()),
+                Excluded(v) => Included(v.clone()),
+                Unbounded => Unbounded,
+            }
+        }
+        if !matches!(start, Unbounded) {
+            complement_segments.push((start, Unbounded));
+        }
+
+        Self {
+            segments: complement_segments,
+        }
+    }
+}
+
+impl<V: Ord> BoundedRange<V> {
+    /// Returns true if the this Range contains the specified value.
+    pub fn contains(&self, v: &V) -> bool {
+        for segment in self.segments.iter() {
+            if match segment {
+                (Unbounded, Unbounded) => true,
+                (Unbounded, Included(end)) => v <= end,
+                (Unbounded, Excluded(end)) => v < end,
+                (Included(start), Unbounded) => v >= start,
+                (Included(start), Included(end)) => v >= start && v <= end,
+                (Included(start), Excluded(end)) => v >= start && v < end,
+                (Excluded(start), Unbounded) => v > start,
+                (Excluded(start), Included(end)) => v > start && v <= end,
+                (Excluded(start), Excluded(end)) => v > start && v < end,
+            } {
+                return true;
+            }
+        }
+        false
+    }
+
+    /// Construct a simple range from anything that impls [RangeBounds] like `v1..v2`.
+    pub fn from_range_bounds<R, IV>(bounds: R) -> Self
+    where
+        R: RangeBounds<IV>,
+        IV: Clone + Into<V>,
```

Comment:
I was confused by why this required clone, I had worked hard to find a way to do this without the extra clone [here](https://github.com/pubgrub-rs/pubgrub/pull/105/files#diff-fbd4e90bd1c07576edba7f752cccda844f7209ef6f79c25b665bbeceb97c629cR94). But this may be micro optimization on my part, `clone` is a much more common bound. Thoughts?

Reply:
Ah yes, I changed that because most types don't implement `From<&'a T> for T`. I felt like that was a bit of a strange requirement to have, especially since you will most likely implement that with a `.clone()` anyway.

You did teach me about `for<'a>`! I had never seen that. 😄